src/charm.py

```diff
@@ -383,35 +391,53 @@ def _get_juju_config_identity_schema_config(self) -> Optional[Tuple[str, Dict]]:
                 logger.warning("Ignoring `identity_schemas` configuration")
                 return None
 
-            try:
-                schemas = json.loads(identity_schemas)
-            except json.JSONDecodeError as e:
-                logger.error(f"identity_schemas is not a valid json: {e}")
-                logger.warning("Ignoring `identity_schemas` configuration")
-                return None
-
             schemas = {
-                schema_id: f"base64://{base64.b64encode(json.dumps(schema).encode()).decode()}"
-                for schema_id, schema in schemas.items()
+                schema_id: f"base64://{base64.b64encode(schema.encode()).decode()}"
+                for schema_id, schema in identity_schemas.items()
             }
             return default_schema_id, schemas
         return None
 
-    def _get_default_identity_schema_config(self) -> Tuple[Optional[str], Optional[Dict]]:
+    def _get_configmap_identity_schema_config(self) -> Optional[Tuple[str, Dict]]:
+        identity_schemas = self.configmap_handler.get_identity_schemas()
+        if not identity_schemas:
+            return None
+
+        default_schema_id = identity_schemas.pop("default.schema")
+        if not default_schema_id:
+            logger.error("`configMap with identity schemas contains no `default.schema`")
+            return None
+
         schemas = {
-            schema_file.stem: f"file://{self._identity_schemas_default_dir_path / schema_file.name}"
-            for schema_file in self._identity_schemas_local_dir_path.glob("*.json")
+            schema_id: f"base64://{base64.b64encode(schema.encode()).decode()}"
+            for schema_id, schema in identity_schemas.items()
         }
+        return default_schema_id, schemas
+
+    def _get_default_identity_schemas(self) -> Dict:
+        schemas = {}
+        for schema_file in self._identity_schemas_local_dir_path.glob("*.json"):
+            with open(schema_file) as f:
+                schema = f.read()
+            schemas[schema_file.stem] = schema
+        return schemas
+
+    def _get_default_identity_schema_config(self) -> Tuple[str, Dict]:
+        schemas = self._get_default_identity_schemas()
         default_schema_id_file = (
             self._identity_schemas_local_dir_path / DEFAULT_SCHEMA_ID_FILE_NAME
         )
         with open(default_schema_id_file) as f:
             default_schema_id = f.read()
         if default_schema_id not in schemas:
             raise RuntimeError(f"Default schema `{default_schema_id}` can't be found")
+        schemas = {
+            schema_id: f"base64://{base64.b64encode(schema.encode()).decode()}"
+            for schema_id, schema in schemas.items()
+        }
         return default_schema_id, schemas
 
-    def _get_identity_schema_config(self) -> Optional[Tuple[str, Dict]]:
+    def _get_identity_schema_config(self) -> Tuple[str, Dict]:
```

Comment:
Saw some common patterns in each identity schema source `get` function. Could be an overkill in the Python world, but I think it would be better when more sources of identity schema show up.

- Extract the identity schema config as a standalone base class. It has a `render` abstract method which can be overridden by different subclasses
- Create different classes that inherit the base class by overriding the `render` method, e.g. `DefaultIdentitySchemaConfig`
- Apply a chain of calling each class's `render` method with a specific order. The chain will return early if some of the handlers return an identity schema

We can control the chain order very easily and extend it easily with more subclasses.

Just a thought. No need to make any changes.

Reply:
It makes sense what you are suggesting, I don't think we will need to add more sources for this in the future, but it would be nice to just separate this logic from the rest of the charm code. If it's OK with you though I would like to do that in a later PR.

Reply:
yeah, definitely don't change further in this PR. We can improve it later.